### PR TITLE
fix: [DEL-3603]: Fixing Ratio for manager and delegate version decoupling

### PR DESCRIPTION
**Pull request by AnupamIO**
JIRA: https://harness.atlassian.net/browse/DEL-3603

Added changes for:
1- Fixing Ratio API after manager and delegate version decoupling
2- Added a new API which will tell what percent of delegate of specific version are connected. (Mostly to be used after deployments)

You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/32162)

### CODE REVIEW

**Review by gaurav-jain1 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32162)* status: 5 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java, line 197 at r1](https://reviewable.io/reviews/harness/harness-core/32162#-N-MqVRw2Ipp9tjkPQR3:-N-MqVRw2Ipp9tjkPQR4:btpcrd0) ([raw file](https://github.com/harness/harness-core/blob/5366ec5dcb928476b210cf5b1a7a44865dfd8e91/400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java#L197)):*
> ```Java
> 
>   @GET
>   @Path("connected-ratio-with-primary")
> ```

Why do we have API at 2 different places ? Can we get rid of one ?



_Code quote:_
```Java
@Path("connected-ratio-with-primary")
```
___
*[400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java, line 220 at r1](https://reviewable.io/reviews/harness/harness-core/32162#-N-N1X8cCeYl8IQeJ1oQ:-N-N1X8cCeYl8IQeJ1oR:b7uuznd) ([raw file](https://github.com/harness/harness-core/blob/5366ec5dcb928476b210cf5b1a7a44865dfd8e91/400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java#L220)):*
> ```Java
>   @PublicApi
>   @Operation(operationId = "getConnectedDelegatesRatio",
>     summary = "Calculates ratio of connected Delegates with version vs total delegates of that version.",
> ```

Calculates ratio of connected Delegates with specific version.



_Code quote:_
```Java
Calculates ratio of connected Delegates with version vs total delegates of that version.
```
___
*[420-delegate-service/src/main/java/software/wings/service/impl/DelegateConnectionDao.java, line 71 at r1](https://reviewable.io/reviews/harness/harness-core/32162#-N-NDNLu6Vd1SbozgAKj:-N-NDNLu6Vd1SbozgAKk:b8fgm3c) ([raw file](https://github.com/harness/harness-core/blob/5366ec5dcb928476b210cf5b1a7a44865dfd8e91/420-delegate-service/src/main/java/software/wings/service/impl/DelegateConnectionDao.java#L71)):*
> ```Java
>   }
> 
>   public long numberOfDelegatePerVersion(String version, String accountId) {
> ```

Consider renaming it to numberOfDelegateConnectionsPerVersion

**Comment by raghuAtWings** [comment]
*[400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java, line 591 at r1](https://reviewable.io/reviews/harness/harness-core/32162#-N-OcCPb1mpLj5ps_yES:-N-OcCPb1mpLj5ps_yET:b-n1nnqe) ([raw file](https://github.com/harness/harness-core/blob/5366ec5dcb928476b210cf5b1a7a44865dfd8e91/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L591)):*
> ```Java
> 
>     // This is a case when no prior delegate version is available for specific account_id
>     if(CollectionUtil.isEmpty(allDelegateVersions)) {
> ```

why would this be the case? shouldn't this be an error case?

**Review by AnupamIO — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32162)* status: 8 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 1 approvals obtained. 

___
*[400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java, line 197 at r1](https://reviewable.io/reviews/harness/harness-core/32162#-N-MqVRw2Ipp9tjkPQR3:-N-PI3zv3qv3drmZ_JNJ:b-g5k7rq) ([raw file](https://github.com/harness/harness-core/blob/5366ec5dcb928476b210cf5b1a7a44865dfd8e91/400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java#L197)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

Why do we have API at 2 different places ? Can we get rid of one ?
</blockquote></details>

As I know we use v1 API from CG UI and v2, v3 based delegate APIs from NG-UI.
___
*[400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java, line 220 at r1](https://reviewable.io/reviews/harness/harness-core/32162#-N-N1X8cCeYl8IQeJ1oQ:-N-PIOc790i4lkuyHzTq:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/5366ec5dcb928476b210cf5b1a7a44865dfd8e91/400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java#L220)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

Calculates ratio of connected Delegates with specific version.
</blockquote></details>

done
___
*[420-delegate-service/src/main/java/software/wings/service/impl/DelegateConnectionDao.java, line 71 at r1](https://reviewable.io/reviews/harness/harness-core/32162#-N-NDNLu6Vd1SbozgAKj:-N-PIVSq4Hy2rVvD9hp_:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/5366ec5dcb928476b210cf5b1a7a44865dfd8e91/420-delegate-service/src/main/java/software/wings/service/impl/DelegateConnectionDao.java#L71)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

Consider renaming it to numberOfDelegateConnectionsPerVersion
</blockquote></details>

done

File: 400-rest/src/main/java/software/wings/resources/DelegateSetupResource.java [hunk 1]
```diff
@@ -186,8 +186,8 @@ public RestResponse<List<String>> getAvailableVersions(@QueryParam("accountId")
   @ExceptionMetered
   @PublicApi
   public RestResponse<Double> getConnectedRatioWithPrimary(
-      @QueryParam("targetVersion") @NotEmpty String targetVersion) {
```
Comment (AnupamIO) [inline 400-rest/src/main/java/software/wings/resources/DelegateSetupResource.java:189]: trigger messagecheck

**Comment by AnupamIO** [comment]
trigger messagecheck

**Comment by AnupamIO** [comment]
trigger ti2

**Review by gaurav-jain1 — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32162)* status: 8 file(s) to review. 3. unresolved discussions. 

___
*[400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java, line 197 at r1](https://reviewable.io/reviews/harness/harness-core/32162#-N-MqVRw2Ipp9tjkPQR3:-N-SZ85g0R_MtxPLJ7ED:bym2j14) ([raw file](https://github.com/harness/harness-core/blob/5366ec5dcb928476b210cf5b1a7a44865dfd8e91/400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java#L197)):*
<details><summary><i>Previously, AnupamIO (Anupam Jaiswal) wrote…</i></summary><blockquote>

As I know we use v1 API from CG UI and v2, v3 based delegate APIs from NG-UI.
</blockquote></details>

This call never happens from UI, it happens in our publish/unpublish workflow.,
Check with Bharat.. if we can get rid of any one of them.. you can do that separately though.
___
*[400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java, line 230 at r3](https://reviewable.io/reviews/harness/harness-core/32162#-N-SZWpoEdPsu0er99E1:-N-SZWpoEdPsu0er99E2:b-b0t2is) ([raw file](https://github.com/harness/harness-core/blob/cae73596563dc60d7a1eaa8da6dd422d1812135f/400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java#L230)):*
> ```Java
>   getConnectedDelegatesRatio(@Parameter(description = "Target version for which the ratio is calculated") @QueryParam(
>                                  "targetVersion") @NotEmpty String targetVersion,
>       @QueryParam("accountId") String accountId) {
> ```

Is the indentation correct here ?



_Code quote:_
```Java
 @QueryParam("accountId") String accountId)
```
___
*[400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java, line 591 at r1](https://reviewable.io/reviews/harness/harness-core/32162#-N-OcCPb1mpLj5ps_yES:-N-SZnkpAa9u2vUHI4-U:bfq5zgo) ([raw file](https://github.com/harness/harness-core/blob/5366ec5dcb928476b210cf5b1a7a44865dfd8e91/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L591)):*
<details><summary><i>Previously, AnupamIO (Anupam Jaiswal) wrote…</i></summary><blockquote>

This can be the case when no delegateConfiguration is there for mentioned accountId, could be there probably when an account is new and doesn't have delegate running. As I know publishing to this collection is done at runtime by a script, delegateConfiguration can be null because of some fault in script.

Not keeping as error to keep the behaviour same as before, this is similar to case with no primary delegates: https://github.com/harness/harness-core/blob/develop/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L591
</blockquote></details>

Yeah, essentially when we haven't publish specific version for that account.

**Review by gaurav-jain1 — COMMENTED** [review]
> Reviewed 2 of 5 files at r1, 1 of 6 files at r2, 4 of 5 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32162)* status: 3. unresolved discussions. 0 of 1 approvals obtained.

**Review by gaurav-jain1 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32162)* status: 2. unresolved discussions.

**Review by gaurav-jain1 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32162)* status: 1. unresolved discussions.

**Review by gaurav-jain1 — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32162)* status: 1. unresolved discussions.

**Review by AnupamIO — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32162)* status: 1. unresolved discussions. 

___
*[400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java, line 230 at r3](https://reviewable.io/reviews/harness/harness-core/32162#-N-SZWpoEdPsu0er99E1:-N-Z7TVWC531lw09Hjf6:b9tnyn1) ([raw file](https://github.com/harness/harness-core/blob/cae73596563dc60d7a1eaa8da6dd422d1812135f/400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java#L230)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

Is the indentation correct here ?
</blockquote></details>

yes, checked.

**Comment by raghuAtWings** [comment]
*[400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java, line 591 at r1](https://reviewable.io/reviews/harness/harness-core/32162#-N-OcCPb1mpLj5ps_yES:-N-ZKnr67bNdl_oAOndy:bib3v1v) ([raw file](https://github.com/harness/harness-core/blob/5366ec5dcb928476b210cf5b1a7a44865dfd8e91/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L591)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

Yeah, essentially when we haven't publish specific version for that account.
</blockquote></details>

that will never be the case right? since software.wings.service.impl.AccountServiceImpl#getDelegateConfiguration returns the global account id's delegate configuration



_Code quote:_
```Java
getDelegateConfiguration
```

**Review by AnupamIO — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32162)* status: 1. unresolved discussions. 

___
*[400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java, line 591 at r1](https://reviewable.io/reviews/harness/harness-core/32162#-N-OcCPb1mpLj5ps_yES:-N-Wjk06CKZCDXHoxfDN:bhqgqbd) ([raw file](https://github.com/harness/harness-core/blob/5366ec5dcb928476b210cf5b1a7a44865dfd8e91/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L591)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

that will never be the case right? since software.wings.service.impl.AccountServiceImpl#getDelegateConfiguration returns the global account id's delegate configuration
</blockquote></details>

that is true, this case will never arise. I'll remove this check as global_account will always have delegateConfiguration and delegate version.

**Review by raghuAtWings — APPROVED** [review]
> Reviewed 2 of 5 files at r1, 1 of 6 files at r2, 3 of 5 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32162)* status: :shipit: complete! ✓

**Review by gaurav-jain1 — APPROVED** [review]
> Reviewed 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32162)* status: :shipit: complete! ✓

**Pull request merged**